### PR TITLE
Incorporate PCI backed kernel interfaces (like SRIOV) into the kernel Mechanism

### PR DESCRIPTION
**Pull request by edwarnicke**

### CODE REVIEW

**Comment by edwarnicke** [comment]
I'm putting this out there as a Draft for discission, @przemeklal @jellonek @rdimitrov @rktidwell : thoughts?

@nickolaev Does this allay some of your previous concerns wrt the kernel interfaces backed by PCI devices?

File: pkg/api/networkservice/mechanisms/kernel/helpers.go [hunk 1]
```diff
@@ -57,3 +62,11 @@ func (m *mechanism) GetNetNSInode() string {
 	}
 	return m.GetParameters()[common.NetNSInodeKey]
 }
+
+// GetPCIAddress returns PCI address of the VF device
+func (m *mechanism) GetPCIAddress() string {
+	if m == nil || m.GetParameters() == nil {
+		return ""
```
Comment (nickolaev) [inline pkg/api/networkservice/mechanisms/kernel/helpers.go]: Can we return a constant here? Like `"KERNEL"`, meaning that the device is not a physical one (incliuding SR-IOV VFs), but rather a pure kernel - tun/tap, veth, macvlan etc.

File: pkg/api/networkservice/mechanisms/kernel/helpers.go [hunk 2]
```diff
@@ -57,3 +62,11 @@ func (m *mechanism) GetNetNSInode() string {
 	}
 	return m.GetParameters()[common.NetNSInodeKey]
 }
+
+// GetPCIAddress returns PCI address of the VF device
```
Comment (nickolaev) [inline pkg/api/networkservice/mechanisms/kernel/helpers.go]: Can we not refer to VF here, we might very well talk of a non-SR-IOV enabled device and still use this mechanism.

**Comment by przemeklal** [comment]
Copying my another comment:

> HW interface is different in many ways (from SW based):
> - it cannot be created or deleted like veth or tap, for example instead you need to do a lookup in sys fs to find the VF with given PCI address, physical interfaces are also not infinite unlike software based ones
> - there's a lot of SRIOV specific functionalities - like the whole administrative/effective MAC thing, the way you configure VLANs, traffic rates, SRIOV specific settings like trust modes, anti-spoofing modes, need to have the PF in UP state to have traffic working on a VF, and so on, and so on - you may not see it currently in the initial PoC but it'll be important at some point
> 
> I agree there's a good chunk of generic options that apply to both and we should figure out a way to not duplicate functionality (e.g. moving links to netns or setting the IP), however I expect there may be even more features that will be SRIOV-only, even in the kernel mode.
> These extra SRIOV only features are going to be relevant in kernel and userspace mode.

Now, if you guys think having a single mechanism in API and two separate SDKs (kernel and sriov one, because we're going to need a separate SDK for the sriov, the overlap between kernel and userspace is too large imo) - then that's fine, but comments from @nickolaev would need to be incorporated.

File: pkg/api/networkservice/mechanisms/kernel/helpers.go (hunk 1 above)
Comment (edwarnicke) [reply pkg/api/networkservice/mechanisms/kernel/helpers.go]: I'd rather have a method like IsPCIDevice() that would properly indicate whether we are or are not a PCIDevice.

File: pkg/api/networkservice/mechanisms/kernel/helpers.go (hunk 2 above)
Comment (edwarnicke) [reply pkg/api/networkservice/mechanisms/kernel/helpers.go]: Very true in this case, definitely will fix.

**Comment by edwarnicke** [comment]
> Copying my another comment:
> 
> > HW interface is different in many ways (from SW based):

The question is: different to *who*.  I would maintain that to the *consuming* Pod, how it interacts with the interface is exactly the same whether its a tapv2, vethpair, or HW backed.  All the things that are different are *not* things it can touch or even see.

> > 
> > * it cannot be created or deleted like veth or tap, for example instead you need to do a lookup in sys fs to find the VF with given PCI address, physical interfaces are also not infinite unlike software based ones
All of which is super super relevant to us as the *platform* and completely irrelevant to the consuming Pod (which can't do any of those things).

> > * there's a lot of SRIOV specific functionalities - like the whole administrative/effective MAC thing, the way you configure VLANs, traffic rates, SRIOV specific settings like trust modes, anti-spoofing modes, need to have the PF in UP state to have traffic working on a VF, and so on, and so on - you may not see it currently in the initial PoC but it'll be important at some point

Are any of these things that the consuming unprivileged Pod can do anything to?  Or even see?

> > 
> > I agree there's a good chunk of generic options that apply to both and we should figure out a way to not duplicate functionality (e.g. moving links to netns or setting the IP), however I expect there may be even more features that will be SRIOV-only, even in the kernel mode.

Which of these features can be manipulated by an unprivileged consuming Pod?

> > These extra SRIOV only features are going to be relevant in kernel and userspace mode.
> 
> Now, if you guys think having a single mechanism in API and two separate SDKs (kernel and sriov one, because we're going to need a separate SDK for the sriov, the overlap between kernel and userspace is too large imo) - then that's fine, but comments from @nickolaev would need to be incorporated.

So my big question here is making sure we talk through the differences you raised enough to get clarity on how much of these SRIOV differences are going to be visible to consuming unprivileged Pods.  I have a (clearly much less informed) view that there aren't, but it helps tremendously to find out now that I'm mistaken rather than later :)

So let's start with one of the features you have in mind, how would an unprivileged Pod see it?

**Comment by przemeklal** [comment]
I see where you're going now :)

From pods' perspective there's no difference, in the "kernel mode" it's just another network interface, it may support more offloading features than e.g. veth or tap but that's about it.

The differences are in how one provisions the interface - these config options need to applied on the host level and a lot of them are SRIOV only. So the difference is from the NSM or more precisely the NSM Forwarder's point of view.

Question - could we have two separate forwarders (SDKs?) serving a single kernel mechanism, either allocating a VF or creating veth interfaces and configuring those? If so, I'm okay with having a common mechanism. If you want to have a PCI address as part of the kernel mechanism request, then it's also fine for me.

File: pkg/api/networkservice/mechanisms/kernel/helpers.go (hunk 1 above)
Comment (rktidwell) [reply pkg/api/networkservice/mechanisms/kernel/helpers.go]: Could we simply return nil here?

**Comment by rktidwell** [comment]
> The differences are in how one provisions the interface - these config options need to applied on the host level and a lot of them are SRIOV only. So the difference is from the NSM or more precisely the NSM Forwarder's point of view.

In my opinion this is an important point. I'm still trying wrap my head around all this, but I do want to be sure we are separating the provisioning of the interface from the presentation of the interface to the pod. I may be missing something but I can't think of a scenario where once provisioned, an interface needs to be treated differently when presented to a pod. As far as I can see, the only differences in logic we will encounter apply to only to provisioning interfaces. Presenting an interface to a pod seems like it is done the same way regardless of the technology backing the interface.

File: pkg/api/networkservice/mechanisms/kernel/helpers.go (hunk 1 above)
Comment (edwarnicke) [reply pkg/api/networkservice/mechanisms/kernel/helpers.go]: @rktidwell In Golang, a type has a 'zero value' ... for pointers its nil, for strings it "".  So returning "" is sort of the string equivalent of returning nil.

All of that said, @nickolaev What do you think of my IsPCIDevice() bool suggestion?

**Comment by edwarnicke** [comment]
> I see where you're going now :)
> 
Oh good!

> From pods' perspective there's no difference, in the "kernel mode" it's just another network interface, it may support more offloading features than e.g. veth or tap but that's about it.
> 
> The differences are in how one provisions the interface - these config options need to applied on the host level and a lot of them are SRIOV only. So the difference is from the NSM or more precisely the NSM Forwarder's point of view.
Totally agree there.  But as consuming Pods are generally going to be unprivileged, tweeking those knobs is going to be up to the Forwarder, not the consuming Pods :)
> 
> Question - could we have two separate forwarders (SDKs?) serving a single kernel mechanism, either allocating a VF or creating veth interfaces and configuring those? If so, I'm okay with having a common mechanism. Edit: I think this can also be done as part of larger all-in-one mechanism as well, if everyone is okay with such approach.
YES!  Exactly.  One of the nice thing about 'Mechanisms' is that you can have more than one Forwarder providing them if need be.  So you could have a SRIOV Forwarder, and if it gets a Request for a PCI Device it doesn't understand how to handle, it just returns an error and hopefully there's some other Forwarder that does understand how to configure that thing.

> 
> If you want to have a PCI address as part of the kernel mechanism request, then it's also fine for me.
As an optional part.  I think there's still an open discussion with @nickolaev about whether to have an IsPCIDevice() bool or HasPCIDevice() bool helper in there... but that's the proposal: if you provide a PIC Address, you are asking for the kernel interface to be backed by the thing there.  If you don't provide the PCI address, we can feel free to back it with a vethpair or tapv2 etc.

**Comment by przemeklal** [comment]
Cool, thanks, it makes perfect sense in my opinion.

File: pkg/api/networkservice/mechanisms/kernel/helpers.go (hunk 1 above)
Comment (nickolaev) [reply pkg/api/networkservice/mechanisms/kernel/helpers.go]: I agree. Let's go with your proposal.

**Pull request merged**